File: November/9/wordsearch2.java
// LeetCode 212: Word Search II
// Explanation:
// 1. Build Trie of words for prefix-based pruning.
// 2. Perform DFS from each board cell to find matches.
// 3. Use backtracking and remove found words to avoid duplicates.
// Time Complexity: O(M * N * 4^L)
// Space Complexity: O(T + M*N)

import java.util.*;

class Solution {
    class TrieNode {
        Map<Character, TrieNode> children = new HashMap<>();
        String word = null;
    }

    public List<String> findWords(char[][] board, String[] words) {
        TrieNode root = new TrieNode();
        for (String w : words) {
            TrieNode node = root;
            for (char c : w.toCharArray())
                node = node.children.computeIfAbsent(c, k -> new TrieNode());
            node.word = w;
        }

        List<String> res = new ArrayList<>();
        int rows = board.length, cols = board[0].length;
        for (int r = 0; r < rows; r++)
            for (int c = 0; c < cols; c++)
                dfs(board, r, c, root, res);
        return res;
    }

    private void dfs(char[][] board, int r, int c, TrieNode node, List<String> res) {
        char ch = board[r][c];
        if (!node.children.containsKey(ch)) return;
        TrieNode next = node.children.get(ch);
        if (next.word != null) {
            res.add(next.word);
            next.word = null;
        }

        board[r][c] = '#';
        int[][] dirs = {{1,0},{-1,0},{0,1},{0,-1}};
        for (int[] d : dirs) {
            int nr = r + d[0], nc = c + d[1];
            if (nr >= 0 && nr < board.length && nc >= 0 && nc < board[0].length && board[nr][nc] != '#')
                dfs(board, nr, nc, next, res);
        }
        board[r][c] = ch;
        if (next.children.isEmpty()) node.children.remove(ch);
    }
}

File: 2025/November/9/wordsearch2.py
# LeetCode 212: Word Search II
# Explanation:
# 1. Build a Trie to efficiently check prefixes and complete words.
# 2. For each board cell, perform DFS with backtracking.
# 3. Prune search when prefix not in Trie; mark visited cells temporarily.
# Time Complexity: O(M * N * 4^L), where L = max word length.
# Space Complexity: O(T + M*N), where T = total chars in words.

class TrieNode:
    def __init__(self):
        self.children = {}
        self.word = None

class Solution:
    def findWords(self, board: list[list[str]], words: list[str]) -> list[str]:
        root = TrieNode()
        for w in words:
            node = root
            for c in w:
                if c not in node.children:
                    node.children[c] = TrieNode()
                node = node.children[c]
            node.word = w
        
        res = []
        rows, cols = len(board), len(board[0])
        
        def dfs(r, c, node):
            char = board[r][c]
            if char not in node.children:
                return
            nxt = node.children[char]
            if nxt.word:
                res.append(nxt.word)
                nxt.word = None  # avoid duplicates

            board[r][c] = '#'
            for dr, dc in [(1,0),(-1,0),(0,1),(0,-1)]:
                nr, nc = r+dr, c+dc
                if 0 <= nr < rows and 0 <= nc < cols and board[nr][nc] != '#':
                    dfs(nr, nc, nxt)
            board[r][c] = char

        for r in range(rows):
            for c in range(cols):
                dfs(r, c, root)
        
        return res

sol = Solution()
board = [["o","a","a","n"],["e","t","a","e"],["i","h","k","r"],["i","f","l","v"]]
print(sol.findWords(board, ["oath","pea","eat","rain"]))  # ["eat","oath"]

board = [["a","b"],["c","d"]]
print(sol.findWords(board, ["abcb"]))  # []
